tests/cpp/test_indexing.cpp

```diff
@@ -232,6 +252,98 @@ class IndexValidator : public kir::IrVisitor {
   GetReference get_ref_;
 };
 
+template <typename GetReference>
+class PredicateIndexValidator : public kir::IrVisitor {
+ public:
+  PredicateIndexValidator(const GpuLower& lower, GetReference&& get_ref)
+      : get_ref_(std::move(get_ref)) {}
+
+  using kir::IrVisitor::dispatch;
+  using kir::IrVisitor::handle;
+
+  void dispatch(Expr* expr) override {
+    if (!ir_utils::isTvOp(expr)) {
+      kir::IrVisitor::dispatch(expr);
+      return;
+    }
+
+    get_ref_.setForLoops(for_loops_);
+
+    if (auto loop_it = std::find_if(
+            for_loops_.begin(),
+            for_loops_.end(),
+            [](ForLoop* fl) {
+              return fl->circularBufferLoopStage() !=
+                  CircularBufferLoopStage::NotApplicable;
+            });
+        loop_it != for_loops_.end()) {
+      auto loop = *loop_it;
+      get_ref_.setCircularBufferInfo(loop->circularBufferLoopStage());
+    }
+
+    auto out_ti = expr->output(0)->as<kir::TensorIndex>();
+
+    NVF_ERROR(!scope_exprs_.empty());
+    auto inline_predicate = dynamic_cast<kir::IfThenElse*>(scope_exprs_.back());
+    NVF_ERROR(
+        inline_predicate != nullptr,
+        "No inline predicate detected: ",
+        expr->toString());
+
+    validate(out_ti, inline_predicate->predicate()->value());
+
+    get_ref_.clearForLoops();
+    get_ref_.clearCircularBufferInfo();
+  }
+
+  void validate(kir::TensorIndex* ti, Val* actual) {
+    TensorView* tv = ti->view();
+    Val* ref = get_ref_.getPredicate(tv);
+    if (ref != nullptr) {
```

Comment:
Should this be `ASSERT_TRUE(ref != nullptr)`? Otherwise if we inserted a predicate that should not be there it would silently pass.

Reply:
This is intentional. The `GetReference` of each test should return a reference Val for any tensor that should be validated. This is convenient since not all tensors need to be verified, and for those `GetReference` can just return nullptr.